cache/cache_reservation_manager.cc

```diff
@@ -0,0 +1,148 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+#include "cache/cache_reservation_manager.h"
+
+#include <cassert>
+#include <cstddef>
+#include <cstring>
+#include <memory>
+
+#include "cache/cache_entry_roles.h"
+#include "rocksdb/cache.h"
+#include "rocksdb/slice.h"
+#include "rocksdb/status.h"
+#include "table/block_based/block_based_table_reader.h"
+#include "util/coding.h"
+
+namespace ROCKSDB_NAMESPACE {
+CacheReservationManager::CacheReservationManager(std::shared_ptr<Cache> cache,
+                                                 bool delayed_decrease)
+    : delayed_decrease_(delayed_decrease), cache_allocated_size_(0) {
+  assert(cache != nullptr);
+  cache_ = cache;
+  std::memset(cache_key_, 0, kCacheKeyPrefixSize + kMaxVarint64Length);
+  EncodeVarint64(cache_key_, cache_->NewId());
+}
+
+CacheReservationManager::~CacheReservationManager() {
+  for (auto* handle : dummy_handles_) {
+    if (handle != nullptr) {
+      cache_->Release(handle, true);
+    }
+  }
+}
+
+template <CacheEntryRole R>
+Status CacheReservationManager::UpdateCacheReservation(
+    std::size_t new_mem_used) {
+  std::size_t cur_cache_allocated_size =
+      cache_allocated_size_.load(std::memory_order_relaxed);
+  if (new_mem_used == cur_cache_allocated_size ||
+      (cur_cache_allocated_size > 0 &&
+       new_mem_used > cur_cache_allocated_size &&
+       (new_mem_used - cur_cache_allocated_size) < kSizeDummyEntry) ||
+      (new_mem_used < cur_cache_allocated_size &&
+       (cur_cache_allocated_size - new_mem_used) < kSizeDummyEntry)) {
```

Comment:
[Note]
While change in memory release behavior is expected since we are revising the memory releasing algorithm, I need to point out that the memory reserving behavior is also changed a bit. 

The previous memory reserving behavior can be described as the following, where "+(new_mem_used)" stands for what happened in [here](https://github.com/facebook/rocksdb/pull/8506/files#diff-72f8c69a5d31088bb7cbda00065f74f10f931a209fe3a5b342326f115738d9c4L107-L125), that is, looping until "new_mem_used > cache_allocated_size_" is no longer true.

Assume we starts with new_mem_used = cache_allocated_size = 0

op no.| new_mem_used|    operation   |  cache_allocated_size after operation
1.         |         128KB         |   +(128KB)   |                  256KB
2.        |          129KB         |   +(129KB)   |                  256KB  (since new_mem_used < cache_allocated_size)
3.        |          257KB         |   +(257KB)  |                   512KB


However, the under the new code, the reserving behavior will become:

Assume we starts with new_mem_used = cache_allocated_size = 0

op no.| new_mem_used|    operation             |  cache_allocated_size after operation
1.         |         128KB         |   update(128KB)   |                  256KB
2.        |          129KB         |   update(129KB)   |                  256KB  (since new_mem_used < cache_allocated_size && (cur_cache_allocated_size - new_mem_used) < kSizeDummyEntry)
3.        |          257KB         |   update(257KB)  |                   256KB (since new_mem_used > cur_cache_allocated_size && (new_mem_used - cur_cache_allocated_size) < kSizeDummyEntry)
       
The reason for such change is that, when we consolidate reserve/free memory reservation into one update memory reservation API, we currently rely on the relative size order of new_mem_used and cache_allocated_size to decide whether to do an increase or a decrease. If we use a strict comparison such as "if new_mem_used > cache_allocated_size, then increase; if new_mem_used < cache_allocated_size, then decrease", we end up doing a decrease for op 2 above in our new behavior. Therefore, I decided to allow a "fuzzy equal condition", that is, abs(new_mem_used - cache_allocated_size) < kSizeDummyEntry can be seen as new_mem_used = cache_allocated_size and we do not need to update any cache reservation under such case.

One exception to the above "fuzzy equal condition" in my own impl is the "cur_cache_allocated_size > 0" in [here](https://github.com/facebook/rocksdb/pull/8506/files#diff-89e4f3dd4e967498e08e77b49f8d290d242f9f1b1fb30f1434480711545917b7R47). The reason for this exception is to adapt to existing test cases related to WBM that assumes trivial memory increase less than 1 dummy entry size (e.g, increase from 0 to 128 KB) always lead to 1 dummy entry insertion. Such test cases includes "TestWriteBufferNoLimitWithCache - DBTest2", "SharedWriteBufferAcrossCFs/2 - DBTestSharedWriteBufferAcrossCFs/DBTestSharedWriteBufferAcrossCFs", "CacheEntryRoleStats - DBBlockCacheTest".

I think such a behavior change in reserving should be fine as CacheRep/WBM:: ReserveMemWithCache/ WriteBufferManager:: FreeMemWithCache are all private implementation details. And the implementation details described in the public document https://github.com/facebook/rocksdb/wiki/Write-Buffer-Manager#implementation is still compatible will the new changes (However, we might need to change the release part). I hope there isn't any users using the public API  WriteBufferManager::dummy_entries_in_cache_usage() to infer back into how the previous cache reservation algorithms work and rely on the previous one.

Do let me know  though if you see any risk in such changes in reservation behavior to our external users that I am not aware of. I can try to investigate (again) to see how to maintain the previous reservation behavior as much as possible while trying to consolidate reserve/free mem into update...

Reply:
> I think it would be simpler to update the reservation bytes to the new_mem_used rounded up to the nearest multiple of kSizeDummyEntry.
Just to make sure I understand, are you suggesting something like the following:
```
rounded_new_mem_used = round_to_nearest(new_mem_used)

if (rounded_new_mem_used == cache_allocated_size_) {
    do nothing 
} else if (rounded_new_mem_used > cache_allocated_size_) {
    increase_to(rounded_new_mem_used) 
} else {       
    decrease_to(rounded_new_mem_used) 
}       

increase_to(rounded_new_mem_used){
      while (cache_allocated_size_ < rounded_new_mem_used && rounded_new_mem_used - cache_allocated_size_ >  kSizeDummyEntry ) {
           insert dummy entry
       }
}
decrease_to(rounded_new_mem_used){
       while (cache_allocated_size_ > rounded_new_mem_used && cache_allocated_size_ - rounded_new_mem_used > kSizeDummyEntry) {
           release dummy entry
       }
}
```

>The main thing I wonder about is if we can simplify how we handle reserving one more or one less dummy entry when the target memory falls in the middle
If I understand your "rounded up to the nearest multiple of kSizeDummyEntry" correctly, the target memory falling in the middle can be rounded up just as 5 rounded up to 10. And we go from there.

Reply:
Sorry for the confusion. I meant to always round up when there is any fraction, like a ceiling function. So 257 rounds up to 512.